waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -379,6 +396,42 @@ validateWebAppBaseDir spec = case maybeBaseDir of
     startsWithSlash ('/' : _) = True
     startsWithSlash _ = False
 
+validateUserNodeVersionRange :: AppSpec -> [ValidationError]
+validateUserNodeVersionRange spec =
+  concat
+    [ checkUserRangeIsInWaspRange,
+      checkUserRangeAllowsOnlyMinorChanges
+    ]
+  where
+    userRange = AS.userNodeVersionRange spec
+    userVersionInterval = SVB.versionBounds userRange
+    waspVersionInterval = SVB.versionBounds $ SV.backwardsCompatibleWith oldestWaspSupportedNodeVersion
+    checkUserRangeIsInWaspRange :: [ValidationError]
+    checkUserRangeIsInWaspRange =
+      if not (userVersionInterval `SVB.isSubintervalOf` waspVersionInterval)
+        then
+          [ GenericValidationError $
+              "Node version range that you specified for your Wasp app ("
+                <> show userRange
+                <> ") has too low lower boundary. It must be >= "
+                <> show oldestWaspSupportedNodeVersion
+          ]
+        else []
+    checkUserRangeAllowsOnlyMinorChanges :: [ValidationError]
+    checkUserRangeAllowsOnlyMinorChanges =
+      maybe
+        [ GenericValidationWarning $
+            "Node version range that you specified for your Wasp app ("
+              <> show userRange
+              <> ") allows major version changes, which is not advised."
+        ]
+        (const [])
+        $ do
+          lowerBoundVersion <- SVB.versionFromBound $ fst userVersionInterval
```

Comment:
I agree @infomiho , this is a bit hard to read. As @sodic said, even better solution is to make code more readable instead of writing comments, so I will go for that. The reason why I didn't make it more readable at start is that it is quite an implementation details and I assumed if somebody really wants to understand this part, they will get them self familiar with the functions used here, but truth is I can relatively easily push some of this logic further away form here and closer to "SemanticVersion" module.

@sodic yes let's improve the code instead! I wouldn't though say there is a lack of semantic names, it is just that these functions are a bit low level for somebody that doesn't have knowledge of SemanticVersion module. But introducing higher level functions, as you suggested, is a nice solution for this.

Ok, done, I moved all of this logic further away, now code here in Valid.hs is really simple, no implementation details.